projects/mci/src/precharge_control.c

```diff
@@ -0,0 +1,145 @@
+  MotorControllerStorage *storage = context;
+  GpioState monitor_state = GPIO_STATE_LOW;
+  gpio_get_state(&storage->settings.precharge_monitor, &monitor_state);
+  if (monitor_state == GPIO_STATE_LOW) {
+    storage->precharge_state = MCI_PRECHARGE_DISCHARGED;
+  } else if (monitor_state == GPIO_STATE_HIGH) {
+    storage->precharge_state = MCI_PRECHARGE_CHARGED;
+    CAN_TRANSMIT_POWER_ON_MAIN_SEQUENCE(NULL, EE_POWER_MAIN_SEQUENCE_PRECHARGE_COMPLETED);
+  }
+}
+
+StatusCode prv_precharge(void *context) {
+  MotorControllerStorage *storage = context;
+  // make sure monitor is off
+  GpioState monitor_state = GPIO_STATE_LOW;
+  gpio_get_state(&storage->settings.precharge_monitor, &monitor_state);
+  if (monitor_state != GPIO_STATE_LOW) {
+    return STATUS_CODE_INTERNAL_ERROR;
+  }
+  // set control to on
+  gpio_set_state(&storage->settings.precharge_control, GPIO_STATE_HIGH);
+  gpio_set_state(&storage->settings.precharge_control2, GPIO_STATE_HIGH);
+  // check control to make sure it's on
+  GpioState control_state = GPIO_STATE_LOW;
+  GpioState control2_state = GPIO_STATE_LOW;
+  gpio_get_state(&storage->settings.precharge_control, &control_state);
+  gpio_get_state(&storage->settings.precharge_control2, &control2_state);
+  if (control_state != GPIO_STATE_HIGH || control2_state != GPIO_STATE_HIGH) {
+    return STATUS_CODE_INTERNAL_ERROR;
+  }
+  return STATUS_CODE_OK;
+}
+
+StatusCode prv_discharge(void *context) {
+  // set precharge_control to off
+  MotorControllerStorage *storage = context;
+  gpio_set_state(&storage->settings.precharge_control, GPIO_STATE_LOW);
+  gpio_set_state(&storage->settings.precharge_control2, GPIO_STATE_LOW);
+  // check precharge_control to make sure it's discharged
+  GpioState control_state = GPIO_STATE_LOW;
+  GpioState control2_state = GPIO_STATE_LOW;
+  gpio_get_state(&storage->settings.precharge_control, &control_state);
+  gpio_get_state(&storage->settings.precharge_control2, &control2_state);
+  if (control_state != GPIO_STATE_LOW || control2_state != GPIO_STATE_LOW) {
+    return STATUS_CODE_INTERNAL_ERROR;
+  }
+  return STATUS_CODE_OK;
+}
+
+StatusCode prv_precharge_rx(const CanMessage *msg, void *context, CanAckStatus *ack_reply) {
+  MotorControllerStorage *storage = context;
+  uint16_t power_sequence = 0;
+  CAN_UNPACK_POWER_ON_MAIN_SEQUENCE(msg, &power_sequence);
+  StatusCode ret = STATUS_CODE_INTERNAL_ERROR;
+  // if it's power on, then prv_precharge
+  if (power_sequence == EE_POWER_MAIN_SEQUENCE_BEGIN_PRECHARGE) {
+    ret = prv_precharge(context);
+  }
+  // if it's power off, then prv_discharge (unimplemented)
+  // if (power_sequence == EE_POWER_MAIN_SEQUENCE_DISCHARGE) {
+  //   ret = prv_discharge(context);
+  // }
+  // reply ack
+  if (ret == STATUS_CODE_OK) {
+    *ack_reply = CAN_ACK_STATUS_OK;
+  } else {
+    *ack_reply = CAN_ACK_STATUS_INVALID;
+  }
+  return ret;
+}
+
+// gpio_init() and gpio_it_init() is required before this is called
```

Comment:
put this at top of the header file: requires gpio and gpio_it to be initialized.

Reply:
take a look at the other header files to see what I mean... e.g. `fsm.h`